### ISSUE TITLE
GTK windows won't receive keyboard events

### ISSUE THREAD

**Issue by hermann-p** [swaywm/sway#4207]
When I start sway, I can operate the WM properly, work from the terminal (`termite`), launch applications, etc.
Still, GTK windows (Firefox/Chromium/Vivaldi) will not receive keyboard events, neither from the laptop's internal keyboard nor from the external one (Cherry). QT apps (qtox, dolphin) seem to work most of the time, though not always; I'm not able to reproduce the bug with those in a reliable way.

I'm running on a DELL xps 15, with free intel and nouveau drivers, Manjaro Linux. Sway is used with a minimal config (default + setting my keyboards' layouts to German).

```
> sway -v                                                                                                                                                                                                    
sway version 1.1-rc1-21-g49258829 (May 31 2019, branch 'master')
```
If there is anything I can do to help you gain more insight, I'll be glad to help.

**Comment by Siborgium** [comment]
I encounter the same problem. From what I noticed, if you close the laptop for a reasonable time (more than ~5s) and then open it again, everything will work fine until the next reboot.

**Comment by schauveau** [comment]
Siborgium, the behavior that you are describing looks suspiciously similar to what I experienced in my ticket https://github.com/swaywm/sway/issues/3843. See the associated ticket in lightdm for more details and for the solution. At that time I also noticed that closing the laptop was a possible solution but I did not mention it because it was not entirely reliable. 
My missing input device problem was not limited to a subset of the applications so it may not apply to the problem described by hermann-p (probably a libinput issue or a bad configuration of X11-input or GTK).

**Comment by Siborgium** [comment]
> My missing input device problem was not limited to a subset of the applications so it may not 

I do not use any display managers, plus the problem is only limited to the certain key combinations in a certain subset of GTK applications. However, I will take a look at the ticket you linked, thanks.

**Comment by hermann-p** [comment]
I found that this depends strongly on the desktop login manager I use. The problem will always appear with `sddm`, often when using `gdm` or `lightdm` (with those, I can log on and off a couple of times until my keys work), and never when I kill any dm and exec `sway` from a tty.
My guess is that there is some problem with the initialisation of drivers; either a race condition or the dm just captures them?

**Comment by RedSoxFan** [comment]
> I found that this depends strongly on the desktop login manager I use. 

From [the wiki](https://github.com/swaywm/sway/wiki#login-managers),
> If you have issues starting sway and you use a login manager, your first step should be disabling the login manager and running sway as described by man 1 sway. If it works, report the bug to your login manager, not to sway.

--

> never when I kill any dm and exec sway from a tty.

Since the issue does not appear when running from a tty, this is not a sway issue.

**Comment by schauveau** [comment]
Lol. "Hello Mister Login Manager! I have a non-reproducible input problem in some Gtk applications while using an obscure Wayland compositor that does not officially support login managers. And I forgot to tell you that the problem also occurs when using other login managers. Can you help me?" 
Good luck with that.
@hermann-p You are screwed :-)

**Comment by schauveau** [comment]
@hermann-p Here is a more constructive remark. Are you sure that this is not a problem with Xwayland? Your X11 applications could be using an incorrect Xkb configuration. 
I am using the following trick to differentiate my X11 applications from my Wayland applications:
```
for_window [shell="xdg"]       title_format '<span color="#00AA00">[%app_id]</span> <span color="#FFAA00">%title</span>'
for_window [shell="xwayland" ] title_format '<span color="#00AA00">[X11]</span> <span color="#FFAA00">%title</span>'
```
Since the problem appears to be specific to the Gtk applications, you should also look at the possible configuration keys that interfere with the wayland configuration. Check for instance if you have some configuration entries in `dconf read /org/gnome/desktop/input-sources/` (see  https://unix.stackexchange.com/questions/279141/how-to-allow-for-caps-lock-keyboard-layout-change )

**Comment by hermann-p** [comment]
@schauveau Thanks for your input. I'm pretty sure that the problem does not relate to Xwayland, as it also affects e.g. `firefox-wayland`, which should run under wayland natively if I'm not mistaken.
I will check out the link you sent me after work.

As a side note, I do not feel that issues originating in the way users log in should be ignored by a window manager. Since repeatedly logging on and off or closing/reopening the laptop make the issues go away, it might be possible that  modifying some init scripts could do the trick.
The majority of users apply login managers. For me, having to manually kill the window manager before starting sway is a PITA and one of two reasons why I don't use sway more often (the other one being that awesomeWM offers a complete taglist per screen while sway distributes a fixed number of tasks among screens; but awesome crashes after ~4-5h when run without a login manager...).
Please do not take this as ill-meant criticism, it is not meant that way, but as a feature request. Supporting login managers could increase your base of users and supporters.

**Comment by emersion** [comment]
>As a side note, I do not feel that issues originating in the way users log in should be ignored by a window manager. Since repeatedly logging on and off or closing/reopening the laptop make the issues go away, it might be possible that modifying some init scripts could do the trick.

Sorry, we don't support login managers.

>The majority of users apply login managers.

I'm not sure about that.

>For me, having to manually kill the window manager before starting sway is a PITA and one of two reasons why I don't use sway more often (the other one being that awesomeWM offers a complete taglist per screen while sway distributes a fixed number of tasks among screens; but awesome crashes after ~4-5h when run without a login manager...).

Then please investigate and fix it.

>Supporting login managers could increase your base of users and supporters.

Sorry, we're not interested in this. We believe it will only bring more bug reports.

**Comment by schauveau** [comment]
@hermann-p If you haven't already done so, you should have a look at my ticket #3843..
There is a race condition between Sway and the VT switch that causes some input devices to be missing. This is not exactly what you are experiencing (because the effect is not limited to a subset of the applications) but I would not be surprised if that was the cause of some of yours problems. The fact that the but disappears with some login manager when logging multiple times is a good indication that there is a race-condition. If so a simple `sleep 1` before starting sway should be enough.

**Comment by schauveau** [comment]
@hermann-p Real men don't use login managers. Real men start sway from a serial line connected to a  VT100 terminal and only using posix /bin/sh commands encrypted in rot13 for improved security :-)

**Comment by ascent12** [comment]
One thing to note is that while we don't "support" login managers, we shouldn't particularly be doing anything that would stop them from working. We're calling all of the logind APIs correctly as far as I know. None of the developers are particularly interested in going out of their way to test any login managers with sway.

I cannot fathom why certain applications wouldn't work with login manager configurations, unless they're relying on truly broken shit like XDG_SESSION_TYPE, in which case, that is a serious application bug.

**Comment by hermann-p** [comment]
@ascent12 I strongly agree with your points. There's a couple of applications (especially flatpack-ed ones) that need delicate startup scripts just to start up on any setup (I'm looking at you, Authenticator).

I think I'll invest some time to tinker around with the session startup scripts, as I also believe that there is either  some interfering setup or some race condition, maybe something of value will fall out of this.

@schauveau I think, real men won't rely on rot13, but encode their commands with elliptic-curve diffie-helman in their head.

**Comment by schauveau** [comment]
@ascent12 I can save you a bit of time in your investigations: The race condition described in #3843 is caused by Sway doing its libinput configuration before the completion of the VT switch initiated by the login manager. Libinput takes care that each seat can only access the input devices while its respective VT is active. Consequently the input device configurations that are started while the old VT is still active are failing and Sway ignores those input devices for the rest of the session.
